sbi-doc PR (Updated)

atishp04 opened the pull request:
I have tried to create the SBI spec as per my knowledge and understanding RISC-V system. There may be some glaring mistakes either in language or technical content. Apologies in advance :).

As my knowledge on memory model is very limited, I have not described the memory ordering related functions. Kindly update.

Please feel free to edit/comment on anything and everything.

atishp04 commented:
ping ?

aswaterman commented on riscv-sbi.md:
````diff
@@ -0,0 +1,113 @@
+```C
+void sbi_send_ipi(const unsigned long *hart_mask)
+```
+Send an inter-processor interrupt to all the harts defined in hart_mask.
````
Suggest adding the following text:

"Interprocessor interrupts manifest at the receiving harts as Supervisor Software Interrupts.

hart_mask is a physical address that points to a bit-vector of harts.  The bit vector is represented as a sequence of unsigned longs whose length equals the number of harts in the system divided by the number of bits in an unsigned long, rounded up to the next integer.

If hart_mask is NULL, an IPI is sent to every hart in the system."

atishp04 replied:
Sure.

aswaterman commented on riscv-sbi.md:
````diff
@@ -0,0 +1,113 @@
+```C
+void sbi_clear_ipi(void)
+```
+Clears any pending IPIs in the hart for which this SBI is invoked.
````
I would say "Clears the pending IPI, if any," to make it clear only one IPI can be pending.

atishp04 replied:
Sure. Thanks for the clarification

aswaterman commented on riscv-sbi.md:
````diff
@@ -0,0 +1,113 @@
+```C
+void sbi_remote_fence_i(const unsigned long *hart_mask)
+```
+Instruct remote harts to execute FENCE.I instruction.
````
"hart_mask is as described in sbi_send_ipi."

atishp04 replied:
ok

aswaterman commented:
Good start!  Do you want to address my feedback now, or in a follow-up PR?

atishp04 commented:
I will update this PR addressing your comments.

palmer-dabbelt commented on riscv-sbi.md:
```diff
@@ -0,0 +1,113 @@
+
+## Introduction
+
+Supervisor Binary Interface(SBI) describes the instructions that can be executed together with a set of SBI calls out to the supervisor execution environment (SEE)
```
missing space

atishp04 replied:
Fixed.

palmer-dabbelt commented on riscv-sbi.md:
```diff
@@ -0,0 +1,113 @@
+Supervisor Binary Interface(SBI) describes the instructions that can be executed together with a set of SBI calls out to the supervisor execution environment (SEE)
+on a given platform. Several instructions that might normally be handled by the
+supervisor operating system (OS) directly are handled via SBI calls to the SEE in
+RISC-V. This provides a cleaner interface for the supervisor OS. Currently,
```
trailing "currently,"

atishp04 replied:
Fixed already.

palmer-dabbelt commented on riscv-sbi.md:
```diff
@@ -0,0 +1,113 @@
+
+## Description
+Each SBI call adheres to a specific calling convention that defines how data is provided as input to or read as output. Individual SBI call signature and its
+purpose is described next.
```
The SBI calling convention matches the syscall calling convention (which in turn is very similar to the function calling convention).

palmer-dabbelt commented on riscv-sbi.md:
```diff
@@ -0,0 +1,113 @@
+Programs the clock for next event after *stime_value* time. This function also
+clears the pending timer interrupt bit. If the supervisor needs to clear the
+timer interrupt without scheduling the next timer event, they may choose to do
+so in timer interrupt handler or any other place in supervisor.
```
I'd just say "by modifying SIE directly".

aswaterman replied:
That's a little misleading, since clearing sie.STIE just masks the timer interrupt, rather than clearing it.

Maybe something more explicit:

"If the supervisor wishes to clear the timer interrupt without scheduling the next timer event, it can either request a timer interrupt infinitely far into the future (i.e., (uint64_t)-1), or it can instead mask the timer interrupt by clearing sie.STIE."

atishp04 replied:
Added.

palmer-dabbelt commented on riscv-sbi.md:
````diff
@@ -0,0 +1,113 @@
+                                unsigned long size,
+                                unsigned long asid)
+```
+//TODO
````
These just look like remote_fence_i

palmer-dabbelt commented on riscv-sbi.md:
````diff
@@ -0,0 +1,113 @@
+```C
+void sbi_console_putchar(int ch)
+```
+Write data present in *ch* to debug console (blocking)
````
If the console doesn't exist then the character is thrown away.

palmer-dabbelt commented on riscv-sbi.md:
````diff
@@ -0,0 +1,113 @@
+```C
+void sbi_shutdown(void)
+```
+Puts all the harts to shut down state from supervisor point of view.
````
This SBI call doesn't return.

atishp04 commented:
Addressed all the feedbacks and updated some content as well. I have updated the same PR to maintain the history. Let me know if you want me to close this and create a new PR.

aswaterman commented:
Thank you, Atish!  Great start on this document.
